### PR TITLE
Use context managers for some file handling in system modules.

### PR DESCRIPTION
**Pull request by atombrella**
##### SUMMARY
Follow-up to #63840 The changes in this PR shouldn't be changing semantics. And it's also much smaller, so hopefully faster to get reviewed and merged.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
system, core

##### ADDITIONAL INFORMATION
Context managers are more Pythonic for dealing with resources, like files or connections.

### CODE REVIEW

File: lib/ansible/modules/system/hostname.py [hunk 1]
```diff
@@ -422,40 +398,33 @@ class OpenRCStrategy(GenericStrategy):
...
                         name = line[10:].strip('"')
                         break
-            except Exception as e:
-                self.module.fail_json(msg="failed to read hostname: %s" %
-                                          to_native(e), exception=traceback.format_exc())
-        finally:
-            f.close()
+        except Exception as e:
+            self.module.fail_json(msg="failed to read hostname: %s" %
+                                      to_native(e), exception=traceback.format_exc())
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:409]: This changes behavior slightly (exception is now caught when error happens during closing), but I think that's ok.

File: lib/ansible/modules/system/hostname.py [hunk 2]
```diff
@@ -422,40 +398,33 @@ class OpenRCStrategy(GenericStrategy):
...
+            with open(self.HOSTNAME_FILE, 'w') as f:
                 f.write('\n'.join(lines) + '\n')
-            except Exception as e:
-                self.module.fail_json(msg="failed to update hostname: %s" %
-                                          to_native(e), exception=traceback.format_exc())
-        finally:
-            f.close()
+        except Exception as e:
+            self.module.fail_json(msg="failed to update hostname: %s" %
+                                      to_native(e), exception=traceback.format_exc())
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:427]: Same here.

File: lib/ansible/modules/system/hostname.py [hunk 3]
```diff
@@ -537,45 +501,39 @@ def get_permanent_hostname(self):
...
                         name = line[10:].strip('"')
                         break
-            except Exception as e:
-                self.module.fail_json(msg="failed to read hostname: %s" %
-                                          to_native(e), exception=traceback.format_exc())
-        finally:
-            f.close()
+        except Exception as e:
+            self.module.fail_json(msg="failed to read hostname: %s" %
+                                      to_native(e), exception=traceback.format_exc())
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:518]: Same here.

File: lib/ansible/modules/system/hostname.py [hunk 4]
```diff
@@ -537,45 +501,39 @@ def get_permanent_hostname(self):
...
+            with open(self.HOSTNAME_FILE, 'w') as f:
                 f.write('\n'.join(lines) + '\n')
-            except Exception as e:
-                self.module.fail_json(msg="failed to update hostname: %s" %
-                                          to_native(e), exception=traceback.format_exc())
-        finally:
-            f.close()
+        except Exception as e:
+            self.module.fail_json(msg="failed to update hostname: %s" %
+                                      to_native(e), exception=traceback.format_exc())
```
Comment (felixfontein) [inline lib/ansible/modules/system/hostname.py:536]: Same here.

**Review by felixfontein — APPROVED** [review]
There are minimal behavior changes, but I think these are fine. The old behavior (exceptions are not caught when they happened during file closing) was probably not intended anyway.

**Comment by atombrella** [comment]
@felixfontein Any reason why this has not been merged yet?

**Comment by felixfontein** [comment]
TBH I somehow assumed this (or parts of it) was core-support as well. Would have also been nice if anyone else (f.ex. maintainers of the involved modules) would have bothered to take a look as well...

**Comment by jborean93** [comment]
Sorry for missing this PR it seems to have slipped through the cracks. We've just come across this PR when going through the old PRs and issues and this one slipped through. We are happy to review a PR against the latest code today to move to context managers, especially since we no longer need old Python support.

**Pull request closed without merging**